csharp/src/Google.Protobuf/Reflection/Dynamic/DynamicMessage.cs

```diff
@@ -0,0 +1,377 @@
+        {
+            fields.WriteTo(output);
+        }
+
+        /// <summary>
+        /// Used to calculate the size of the serialized message.
+        /// </summary>
+        /// <returns></returns>
+        public int CalculateSize()
+        {
+            int size = memoizedSize;
+            if (size != -1)
+            {
+                return size;
+            }
+            size = fields.GetSerializedSize();
+            memoizedSize = size;
+            return size;
+        }
+
+        /// <summary>
+        /// Parse data as a message of the given type and return it.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        public static DynamicMessage ParseFrom(MessageDescriptor type, ByteString data)
+        {
+            Builder builder = NewBuilder(type);
+            builder.MergeFrom(data);
+            return builder.BuildParsed();
+        }
+
+        /// <summary>
+        /// Construct a IMessage.Builder for the given type.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <returns></returns>
+        public static Builder NewBuilder(MessageDescriptor type)
+        {
+            return new Builder(type);
+        }
+
+        /// <summary>
+        /// Returns the value of the field, represented by field descriptor
+        /// </summary>
+        /// <param name="fd"></param>
+        /// <returns></returns>
+        public object GetField(FieldDescriptor fd)
+        {
+            return fields.GetField(fd);
+        }
+
+        private DynamicMessage(MessageDescriptor type, FieldSet fields)
+        {
+            this.type = type;
+            this.fields = fields;
+        }
+
+
+        /// <summary>
+        /// Builder for dynamic messages. Instances are created with DynamicMessage.CreateBuilder..
+        /// </summary>
+        public sealed partial class Builder : IMessage
+        {
+
+            private readonly MessageDescriptor type;
+            private FieldSet fields;
+
+            internal Builder(MessageDescriptor type)
+            {
+                this.type = type;
+                this.fields = FieldSet.CreateInstance();
+            }
+
+            private bool IsInitialized
+            {
+                get { return fields.IsInitializedWithRespectTo(type.Fields.ByJsonName()); }
+            }
+
+            MessageDescriptor IMessage.Descriptor => throw new NotImplementedException();
+
+            /// <summary>
+            /// Creates the DynamicMessage
+            /// </summary>
+            /// <returns></returns>
+            /// <exception cref="Exception"></exception>
+            public DynamicMessage Build()
+            {
+                if (fields != null && !IsInitialized)
+                {
+                    throw new Exception(String.Format("Message {0} is missing required fields", new DynamicMessage(type, fields).GetType()));
```

Comment:
Prefer interpolated strings instead of String.Format()

Reply:
Also not sure what you're trying to accomplish with "new DynamicMessage(...).GetType()", the class is not generic, you know the type statically right? You could just go typeof(DynamicMessage)

Reply:
Updated accordingly